### PR TITLE
[7.x] Add include_data_streams flag for authorization

### PR DESCRIPTION
**Pull request by danhermann**
Most of the work around the addition of an `includeDataStreams` flag was done in #58381. The remainder of this PR adds the flag to the appropriate request classes, updates `IndexNameExpressionResolver` to rely on the flag rather than a separate boolean flag, and updates tests.

~~This PR adds an `includeDataStreams` flag to the `IndicesRequest` interface so that it is available when `AuthorizationEngine::loadAuthorizedIndices` is called. This is necessary to avoid different behaviors when security is enabled and disabled.~~

~~The disparate behavior is described in the referenced issue. When a `GET */_alias` request (or any other request that does not operate on data streams) is sent without security enabled, any data streams present are ignored by the `IndexNameExpressionResolver` class and the correct HTTP 200 response is returned.~~

~~When security _is_ enabled, the star in `GET */_alias` is first resolved to all authorized indices, aliases, and data streams. Because the `_alias` endpoint does not understand data streams, the first data stream it encounters among the list of authorized indices is treated as an alias and an incorrect 404 for "unknown alias" is returned.~~

~~This PR changes `AuthorizationService` to selectively exclude (in the case of most requests) or include (and the case of `ResolveIndexAction.Request` and other requests that opt in) data streams during the authorized index resolution phase.~~

~~Especially interested in feedback from @elastic/es-security.~~

Fixes https://github.com/elastic/elasticsearch/issues/57712

Relates to #53100 

Backport of #58154

### CODE REVIEW

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/oss-distro-docs

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 1]
```diff
@@ -207,7 +207,7 @@ public ClusterState execute(ClusterState currentState) {
                     throw new ConcurrentSnapshotExecutionException(repositoryName, snapshotName, " a snapshot is already running");
                 }
                 // Store newSnapshot here to be processed in clusterStateProcessed
-                List<String> indices = Arrays.asList(indexNameExpressionResolver.concreteIndexNames(currentState, request));
+                indices = Arrays.asList(indexNameExpressionResolver.concreteIndexNames(currentState, request));
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:210]: 😱 , but luckily assertions tripped!

File: server/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java [hunk 2]
```diff
@@ -1702,26 +1717,26 @@ public void testIndicesAliasesRequestTargetDataStreams() {
...
         }
 
-        {
-            IndicesAliasesRequest.AliasActions aliasActions = IndicesAliasesRequest.AliasActions.add().index(dataStreamName)
-                .alias("my-data");
-            String[] indices = indexNameExpressionResolver.concreteIndexNames(state, aliasActions, true);
-            assertEquals(1, indices.length);
-            assertEquals(backingIndex.getIndex().getName(), indices[0]);
-        }
+        //{
```
Comment (martijnvg) [inline server/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java:1733]: why is this commented out?

Comment (danhermann) [reply server/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java:1733]: I missed that in an earlier commit. The backport of https://github.com/elastic/elasticsearch/pull/57900/ will un-comment it.

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/2

**Pull request merged**